keps/sig-scheduling/4328-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,359 @@
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `affinityEviction` controller to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+
+- This KEP does not aim to Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type. 
+- This KEP does not aim to Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAntiAffinity** type. 
+  
+## Proposal
+- Implement the "RequiredDuringSchedulingRequiredDuringExecution" node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### Optimizing Workloads Migrating
+As a Data Scientist, I want to ensure that my GPU-intensive ml workloads are always on nodes with correct type of GPUs during both scheduling and execution. 
+So I labeled one node with "gpu=g1". I added some more powerful nodes, and now I want to move all pods requiring gpu to this new node. 
+With this feature, I can simply remove the "gpu=g1" from the first node and adds it to the new node, so I can move the target pods and keep the remaining pods intact.
+So far descheduler has the ability to achieve this goal. However, bringing a new component to the system introduces both a lot of uncertainty and additional learning costs. So as a user, I hope that kubernetes can natively support this feature so that users can do it without spending extra learning costs.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead to the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+However, the impact won't be significant due to the limitation on the number of pods per node.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeAffinity struct {
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector `json:"requiredDuringSchedulingRequiredDuringExecution,omitempty"`
+}
+```
+
+Add a controller called `affinityEviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node, checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if the selector still met to the new node label. If not, evict the pod.
+- It will add a DisruptionTarget condition with corresponding reason and message for the pod.
+- It will use client-go to delete the pod directly. 
+- It does not respect your configured PodDisruptionBudget.
```

Comment:
Why would this not respect `PodDisruptionBudget`?

Reply:
You can refer to this comment: https://github.com/kubernetes/enhancements/pull/4329#discussion_r1424441363
I think in our user cases, when we use `RequiredDuringExecution`, it means we're ok with this service be disrupted.

Reply:
For instance, if a node label(gpu related) is removed in node A and added in node B, but node B has no enough CPU to run the pod at the time and after eviction, there is no node to schedule for this pod. It would make the application down in such a case. Of course, this sounds like an acceptable case that may make user suffer.

Reply:
Yes, that's why I think users should should choose to use this strict affinity carefully. 
They should enable this affinity when they're sure the pod cannot run on a node that does not meet the affinity or they can tolerate the service be disrupted. 
If the pod definitely can't run on the node which doesn't met the affinity, then it doesn't matter if it will got rescheduled soon(it will result into error if it didn't got evicted anyway).
This is also match the behavior of the taint-based-eviction.